issue#7845: [Shopify] Implement INSERT for the Orders table

marcoscandroglio opened the pull request:
## Description

This change adds the insert and create_orders methods in the OrdersTable class in shopify_tables.py allowing the shopify handler to insert orders into a shopify store orders table.  I will link screenshots of query tests and add documentation changes as required below in the coming days.

Fixes #7845 

### Creating a Shopify Database

**1. Testing CREATE DATABASE**
<img width="1280" alt="create_database" src="https://github.com/mindsdb/mindsdb/assets/61676830/572b1615-59b4-4e5d-97f9-93882349fe6e">

### INSERTing INTO Orders Table

**1. Testing INSERT line_items**
<img width="1279" alt="insert_line_items" src="https://github.com/mindsdb/mindsdb/assets/61676830/cc1e3423-550b-4c51-b539-644e32b917c2">


**2. Testing INSERT All Available Columns**
<img width="1279" alt="insert_all_columns" src="https://github.com/mindsdb/mindsdb/assets/61676830/1bb47d44-ba5d-44de-bf6b-807076d46342">


### Viewing Inserted Orders

**1. SELECT * FROM shopify_datasource.orders**
<img width="1279" alt="orders_table" src="https://github.com/mindsdb/mindsdb/assets/61676830/9b4904c7-e32c-4cc7-a6c4-36bff6e005ac">

**2. Orders As Seen On Shopify Store Orders Dashboard**
<img width="1280" alt="shopify_orders" src="https://github.com/mindsdb/mindsdb/assets/61676830/0185d6fb-bf3d-4935-a38b-eed2b74eb836">


## Type of change

(Please delete options that are not relevant)

- [x] ⚡ New feature (non-breaking change which adds functionality)
- [x] 📄 This change requires a documentation update

## Verification Process

To ensure the changes are working as expected:

 - [x]   Test Location: [https://github.com/marcoscandroglio/mindsdb-tests/blob/main/shopify/shopify_insert_orders.md](url) and copied into the Description section above.
 - [ ]   Verification Steps: Outline the steps or queries needed to validate the change. Include any data, configurations, or actions required to reproduce or see the new functionality.

## Additional Media:

- [ ] I have attached a brief loom video or screenshots showcasing the new functionality or change.

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have appropriately commented on my code, especially in complex areas.
- [x] Necessary documentation updates are either made or tracked in issues.
- [ ] Relevant unit and integration tests are updated or added.

MinuraPunchihewa reviewed (commented):
Hey @marcoscandroglio,
The code looks good!
If you can add the screenshots along with updates to the README based on your change, I think we should be good to go.

marcoscandroglio commented:
@MinuraPunchihewa I have added the screenshots of the queries for this function here for your review: [https://github.com/marcoscandroglio/mindsdb-tests/blob/main/shopify/shopify_insert_orders.md](url).  I am working on the README update and will open a pull request once that is done.

marcoscandroglio commented:
@MinuraPunchihewa I have pushed the README updates for your review and comment.

MinuraPunchihewa reviewed (commented):
Hey @marcoscandroglio,
The README looks great! I cannot seem to view the screenshot though, probably because I do not have access to your fork. Can we please add it to the description in the PR?

marcoscandroglio commented:
@MinuraPunchihewa I have copied the contents of my report with the screenshots into the Description section above.

MinuraPunchihewa reviewed (commented):
Hey @marcoscandroglio,
I have a couple of questions,
- How is it possible to run `INSERT` statements without including an ID? Is it auto-generated?
- Would it be possible for you to take a look at structuring the data in the `line_items` column? To be able to enter each aspect of the line item as a different column? Since this goes beyond supporting `INSERT`, I think we can even take this up as a separate issue.

marcoscandroglio commented:
@MinuraPunchihewa ,


_-     How is it possible to run INSERT statements without including an ID? Is it auto-generated?_

The ID is auto-generated by the Shopify Store.  Looking at the Order documentation page I see that 'id' is a read-only integer.  The 'order_number' (the ones starting at 1001) are also read-only integers.

_-     Would it be possible for you to take a look at structuring the data in the line_items column? To be able to enter each aspect of the line item as a different column? Since this goes beyond supporting INSERT, I think we can even take this up as a separate issue._

This is something I have been thinking about also and my latest commit has been edited to only use each component of line_items as a column and then build the line_items object before sending the entire order through the API.

marcoscandroglio commented:
Here are updated screenshots showing the working changes:

### INSERTing INTO Orders Table

**1. Testing INSERT title, price**
![minimum_columns](https://github.com/mindsdb/mindsdb/assets/61676830/3ed43043-c2ed-4ea4-9ac9-c6ea0be6492a)

**2. Testing INSERT All Available Columns**
![all_columns](https://github.com/mindsdb/mindsdb/assets/61676830/69de5a15-60aa-4d2e-92a7-a2e1bf2db990)

### Viewing Inserted Orders

**1. Orders As Seen On Shopify Store Orders Dashboard**
![shopify_orders](https://github.com/mindsdb/mindsdb/assets/61676830/3119eace-2db8-401d-8f60-b05e24aad035)

marcoscandroglio commented:
@MinuraPunchihewa , it seems I made a mistake earlier when I attempted to rebase and force push.  I think the staging branch in my fork was not fully updated before I proceeded.  It's unclear to me how I can properly resolve this.  I apologize for the inconvenience and would appreciate some guidance with this matter.  Please let me know if there is anything I can do to help.

marcoscandroglio commented:
@MinuraPunchihewa , the changes as shown in my most recent commit are representative of what I intended to do.

MinuraPunchihewa reviewed (commented):
Hey @marcoscandroglio,
The INSERT command looks great! My apologies, but one thing that I did forget to mention is that since we have updated the columns supported by INSERT, we need to ensure that SELECT and UPDATE allows the same columns as well. It will be a little inconsistent across the table otherwise. Is this something you can take a look at?

marcoscandroglio commented:
@MinuraPunchihewa , 
Great! I'll take a look at making the columns consistent between these operations and push any relevant changes.

MinuraPunchihewa commented:
@marcoscandroglio Thank you!

marcoscandroglio commented:
@MinuraPunchihewa , I was able to take a look at column consistency between INSERT, SELECT, and UPDATE and the following are my findings:

### INSERT statement with all supported columns and fields (line_item sub-components):
![INSERT_all_fields_3](https://github.com/mindsdb/mindsdb/assets/61676830/d1192107-0d83-4f04-8ebe-b50ff8a1d777)

### SELECT statement selecting INSERTed columns:
![SELECT_all_INSERT_columns](https://github.com/mindsdb/mindsdb/assets/61676830/67d12259-237b-40fc-afc3-7b1e6ccbb8ac)

### Failed UPDATE statement:
I was not able to successfully run an UPDATE statement with the data in my orders table (I double checked that the id I used was correct).  The error message I'm getting (in screenshot below) is suggesting that it is expecting the `shipping_address` object which is an array similar to the `line_items` object.  I had to work through a similar error when developing the INSERT function.  I'll continue to take a look at this and see if there is any insight from having developed the INSERT function that I can transfer to solving this problem.

![UPDATE_statement](https://github.com/mindsdb/mindsdb/assets/61676830/be40cec6-201f-462e-b968-5829db0cfe6f)

### Regarding supported columns for UPDATE:
After looking into the docs for the Shopify REST API regarding updating orders I found that the update function can only be used on a limited number of fields and cannot be used to update order item information.  From the docs:

_"This operation allows for updating properties of an order including `buyer_accepts_marketing`, `email`, `phone`, `note`, `tags`, `metafields` and `shipping_address_attributes`. It is not for editing the items of an order."_
link to doc source: [https://shopify.dev/docs/api/admin-rest/2024-01/resources/order#put-orders-order-id](url)

MinuraPunchihewa commented:
Hey @marcoscandroglio,
Sorry for the late response, but I think this is fine. Let's just add support for the fields that the API allows and mention it in the README + documentation.
Thank you so much for all your hard work.

marcoscandroglio commented:
@MinuraPunchihewa, I have added support for all the fields where the API allows data to be inserted successfully through this operation. I have also updated the README and the documentation files. Note that fields requiring other features that have not yet been implemented and fields slated for deprecation have not been added. Some of the API objects have fields with identical names, so I appended brief subscripts to all fields that belong to an API object (for example, billing_address has a field called address1_ba and  shipping_address has a field called address1_sa). 

Thank you in advance for taking the time to review this and please let me know if you have any comments or questions!

## INSERT statement with all supported columns and fields
![1_successful_query](https://github.com/mindsdb/mindsdb/assets/61676830/01b4dcc4-8754-498b-8530-c39ba6c7ef5e)

## Result of INSERT query (21 screenshots)
![1_select](https://github.com/mindsdb/mindsdb/assets/61676830/fdea1f8f-9035-45fe-88b4-449743001a1f)
![2_select](https://github.com/mindsdb/mindsdb/assets/61676830/ebcd4888-248e-4f40-869b-7d198b591d34)
![3_select](https://github.com/mindsdb/mindsdb/assets/61676830/5f90abfe-e525-43f0-8ef5-870adb10945b)
![4_select](https://github.com/mindsdb/mindsdb/assets/61676830/0955988d-e9fa-4ff5-bd42-9dca0cb24874)
![5_select](https://github.com/mindsdb/mindsdb/assets/61676830/282e7baf-71b3-458b-9480-2eb39c098a33)
![6_select](https://github.com/mindsdb/mindsdb/assets/61676830/d68b2504-09f9-443b-947b-503b1736056b)
![7_select](https://github.com/mindsdb/mindsdb/assets/61676830/7d58f29a-f689-4487-a413-8b4e13f770fd)
![8_select](https://github.com/mindsdb/mindsdb/assets/61676830/fcce7bdd-b453-479f-8db1-ce5f66b29d1e)
![9_select](https://github.com/mindsdb/mindsdb/assets/61676830/67d11e3e-89e5-488c-956d-517790032d31)
![10_select](https://github.com/mindsdb/mindsdb/assets/61676830/0e7eac7e-78d7-4117-b513-bc6a9e5dd6c7)
![11_select](https://github.com/mindsdb/mindsdb/assets/61676830/7a72c030-53d3-4d74-8f64-d17e1862b6ac)
![12_select](https://github.com/mindsdb/mindsdb/assets/61676830/edbe236e-0f04-4a74-8b8f-0dec8b251199)
![13_select](https://github.com/mindsdb/mindsdb/assets/61676830/25cb6246-f4fc-4704-ad44-acbc504449b3)
![14_select](https://github.com/mindsdb/mindsdb/assets/61676830/d629b606-7152-49fe-bc87-b15c52b4b5d6)
![15_select](https://github.com/mindsdb/mindsdb/assets/61676830/811d3478-4b93-495d-8ace-18e8a3b28d0a)
![16_select](https://github.com/mindsdb/mindsdb/assets/61676830/a5e501dc-2c1c-46de-b0ab-8b8faa8d9f60)
![17_select](https://github.com/mindsdb/mindsdb/assets/61676830/2c1b4503-da66-40d9-8c87-10ec3eed4c0b)
![18_select](https://github.com/mindsdb/mindsdb/assets/61676830/af3e7989-c39b-4b1f-99f1-9d5282e87533)
![19_select](https://github.com/mindsdb/mindsdb/assets/61676830/e0f5a3e8-34fb-49ac-a10b-531c24c2c7cf)
![20_select](https://github.com/mindsdb/mindsdb/assets/61676830/8012536a-5ce8-4468-897a-7f124d2fa769)
![21_select](https://github.com/mindsdb/mindsdb/assets/61676830/59463799-6ea2-4ccb-836d-4285af643350)

marcoscandroglio commented:
@MinuraPunchihewa Please let me know if you have any comments or feedback regarding my latest work on this issue and if there are any actions you need me to take.

MinuraPunchihewa commented:
> @MinuraPunchihewa Please let me know if you have any comments or feedback regarding my latest work on this issue and if there are any actions you need me to take.

@marcoscandroglio Apologies for the delay. I will definitely review this again this week.

marcoscandroglio commented on mindsdb/integrations/handlers/shopify_handler/shopify_tables.py:
```diff
@@ -398,6 +399,55 @@ def select(self, query: ast.Select) -> pd.DataFrame:
+            all_mandatory=False
+        )
+        order_data = insert_statement_parser.parse_query()
+        self.create_orders(order_data)
```
@aayushshah07 , that is a sound suggestion. I will aim to have this updated by the end of this weekend since I will be away from my computer for the next day or so. Out of curiosity, will you be handling future reviews of this pull request?

marcoscandroglio replied:
Upon further inspection of the rest of the file (and a few other files from other handlers similar to this one), it is not standard to make functions analogous to the create_order function (e.g. update_orders function, delete_orders function) protected or private. I'd be interested to hear @MinuraPunchihewa 's thoughts on this. If functions like these should be protected or private it may be a good idea to track these changes in separate issues.

ZoranPandovski commented:
@MinuraPunchihewa Can you please review

MinuraPunchihewa commented:
Hey @marcoscandroglio,
Merging this now! Thank you for your hard work.
